### ISSUE TITLE
Permission Error in jsonrpcserver 4.1.0

### ISSUE THREAD

**Issue by nicola-lunghi** [bcb/jsonrpcserver#111]
Hi
there's a permission issue with the package published on pypi

```
/tmp/jsonrpcserver-4.1.0$ find . -type f -exec ls -lha {} \;
-rw-r--r-- 1 pippo pippo 583 Nov 15 00:09 ./README.md
-rw-r--r-- 1 pippo pippo 4.4K Jan  6 02:07 ./CHANGELOG.md
-rw-r--r-- 1 pippo pippo 1.3K Jan  6 02:07 ./setup.py
-rw------- 1 pippo pippo 73 Sep 10 01:56 ./MANIFEST.in
-rw-r--r-- 1 pippo pippo 3.0K Jan  6 00:06 ./jsonrpcserver/methods.py
-rw-r--r-- 1 pippo pippo 151 Jan  6 00:06 ./jsonrpcserver/__init__.py
-rw-r--r-- 1 pippo pippo 3.2K Jan  6 00:06 ./jsonrpcserver/async_dispatcher.py
-rw------- 1 pippo pippo 921 Jan 30  2019 ./jsonrpcserver/server.py
-rw------- 1 pippo pippo 4.1K Apr 13  2019 ./jsonrpcserver/request.py
-rw------- 1 pippo pippo 2.2K Jan 30  2019 ./jsonrpcserver/log.py
-rw-r--r-- 1 pippo pippo 9.2K Jan  6 02:07 ./jsonrpcserver/response.py
-rw------- 1 pippo pippo 1.3K Jan 30  2019 ./jsonrpcserver/request-schema.json
-rw------- 1 pippo pippo 2.1K Jan 30  2019 ./jsonrpcserver/status.py
-rw-r--r-- 1 pippo pippo 891 Jan  6 00:06 ./jsonrpcserver/exceptions.py
-rw-r--r-- 1 pippo pippo 9.9K Jan  6 02:07 ./jsonrpcserver/dispatcher.py
-rw-r--r-- 1 pippo pippo 38 Jan  6 02:19 ./setup.cfg
-rw-r--r-- 1 pippo pippo 1 Jan  6 00:49 ./jsonrpcserver.egg-info/not-zip-safe
-rw------- 1 pippo pippo 14 Jan  6 02:19 ./jsonrpcserver.egg-info/top_level.txt
-rw------- 1 pippo pippo 566 Jan  6 02:19 ./jsonrpcserver.egg-info/SOURCES.txt
-rw------- 1 pippo pippo 127 Jan  6 02:19 ./jsonrpcserver.egg-info/requires.txt
-rw------- 1 pippo pippo 1.3K Jan  6 02:19 ./jsonrpcserver.egg-info/PKG-INFO
-rw------- 1 pippo pippo 1 Jan  6 02:19 ./jsonrpcserver.egg-info/dependency_links.txt
-rw-r--r-- 1 pippo pippo 1.3K Jan  6 02:19 ./PKG-INFO
-rw------- 1 pippo pippo 1.1K Sep 19  2017 ./LICENSE
```
with the old version
```
:/tmp/jsonrpcserver-4.1.0$ find ../jsonrpcserver-4.0.5/ -type f -exec ls -lha {} \;
-rw-r--r-- 1 pippo pippo 940 Jan 30  2019 ../jsonrpcserver-4.0.5/README.md
-rw-r--r-- 1 pippo pippo 3.5K Sep 10 01:56 ../jsonrpcserver-4.0.5/CHANGELOG.md
-rw-r--r-- 1 pippo pippo 1.1K Sep 10 01:56 ../jsonrpcserver-4.0.5/setup.py
-rw-r--r-- 1 pippo pippo 73 Sep 10 01:56 ../jsonrpcserver-4.0.5/MANIFEST.in
-rw-r--r-- 1 pippo pippo 2.3K Jun 21  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/methods.py
-rw-r--r-- 1 pippo pippo 151 Jan 30  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/__init__.py
-rw-r--r-- 1 pippo pippo 3.2K Apr 13  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/async_dispatcher.py
-rw-r--r-- 1 pippo pippo 921 Jan 30  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/server.py
-rw-r--r-- 1 pippo pippo 4.1K Apr 13  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/request.py
-rw-r--r-- 1 pippo pippo 2.2K Jan 30  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/log.py
-rw-r--r-- 1 pippo pippo 8.9K Jan 30  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/response.py
-rw-r--r-- 1 pippo pippo 1.3K Jan 30  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/request-schema.json
-rw-r--r-- 1 pippo pippo 2.1K Jan 30  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/status.py
-rw-r--r-- 1 pippo pippo 9.5K Jun 22  2019 ../jsonrpcserver-4.0.5/jsonrpcserver/dispatcher.py
-rw-r--r-- 1 pippo pippo 38 Sep 10 01:57 ../jsonrpcserver-4.0.5/setup.cfg
-rw-r--r-- 1 pippo pippo 14 Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/top_level.txt
-rw-r--r-- 1 pippo pippo 502 Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/SOURCES.txt
-rw-r--r-- 1 pippo pippo 127 Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/requires.txt
-rw-r--r-- 1 pippo pippo 1.7K Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/PKG-INFO
-rw-r--r-- 1 pippo pippo 1 Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/dependency_links.txt
-rw-r--r-- 1 pippo pippo 1.7K Sep 10 01:57 ../jsonrpcserver-4.0.5/PKG-INFO
-rw-r--r-- 1 pippo pippo 1.1K Sep 19  2017 ../jsonrpcserver-4.0.5/LICENSE

```

This causes import error in particular for 
```
-rw------- 1 pippo pippo 127 Jan  6 02:19 ./jsonrpcserver.egg-info/requires.txt
-rw------- 1 pippo pippo 1.3K Jan  6 02:19 ./jsonrpcserver.egg-info/PKG-INFO
```
vs
```
-rw-r--r-- 1 pippo pippo 127 Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/requires.txt
-rw-r--r-- 1 pippo pippo 1.7K Sep 10 01:57 ../jsonrpcserver-4.0.5/jsonrpcserver.egg-info/PKG-INFO
```

can you fix it?
thanks
Nick

-> chmod 0644

_Originally posted by @nicola-lunghi in https://github.com/bcb/jsonrpcserver/issues/106#issuecomment-571632418_

**Comment by nicola-lunghi** [comment]
here is how to reproduce in ubuntu
```
mkdir /tmp/test
cd /tmp/test
wget https://files.pythonhosted.org/packages/34/88/f224dcf47cbbbbea001c1bc981ba47dd39dc0ecd0cc98b7956d1839ade16/jsonrpcserver-4.1.0.tar.gz
tar xzf jsonrpcserver-4.1.0.tar.gz
cd jsonrpcserver-4.1.0/
sudo python3 setup.py build install
```
and then
```
python3 -c 'from jsonrpcserver.exceptions import ApiError'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0-py3.7.egg/jsonrpcserver/__init__.py", line 1, in <module>
    from .async_dispatcher import dispatch as async_dispatch
  File "/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0-py3.7.egg/jsonrpcserver/async_dispatcher.py", line 11, in <module>
    from .dispatcher import (
  File "/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0-py3.7.egg/jsonrpcserver/dispatcher.py", line 43, in <module>
    schema = deserialize(resource_string(__name__, "request-schema.json"))
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1157, in resource_string
    self, resource_name
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1401, in get_resource_string
    return self._get(self._fn(self.module_path, resource_name))
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1587, in _get
    with open(path, 'rb') as stream:
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0-py3.7.egg/jsonrpcserver/request-schema.json'
```

```
tree -pu /usr/local/lib/python3.7/dist-packages/jsonrpcserver*
/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0-py3.7.egg
├── [drwxr-sr-x root    ]  EGG-INFO
│   ├── [-rw------- root    ]  dependency_links.txt
│   ├── [-rw-r--r-- root    ]  not-zip-safe
│   ├── [-rw------- root    ]  PKG-INFO
│   ├── [-rw------- root    ]  requires.txt
│   ├── [-rw------- root    ]  SOURCES.txt
│   └── [-rw------- root    ]  top_level.txt
└── [drwxr-sr-x root    ]  jsonrpcserver
    ├── [-rw-r--r-- root    ]  async_dispatcher.py
    ├── [-rw-r--r-- root    ]  dispatcher.py
    ├── [-rw-r--r-- root    ]  exceptions.py
    ├── [-rw-r--r-- root    ]  __init__.py
    ├── [-rw-r--r-- root    ]  log.py
    ├── [-rw-r--r-- root    ]  methods.py
    ├── [drwxr-sr-x root    ]  __pycache__
    │   ├── [-rw-r--r-- root    ]  async_dispatcher.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  dispatcher.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  exceptions.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  __init__.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  log.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  methods.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  request.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  response.cpython-37.pyc
    │   ├── [-rw-r--r-- root    ]  server.cpython-37.pyc
    │   └── [-rw-r--r-- root    ]  status.cpython-37.pyc
    ├── [-rw-r--r-- root    ]  request.py
    ├── [-rw------- root    ]  request-schema.json
    ├── [-rw-r--r-- root    ]  response.py
    ├── [-rw-r--r-- root    ]  server.py
    └── [-rw-r--r-- root    ]  status.py

3 directories, 27 files
```

while if I install with pip is fine
```
sudo pip3 uninstall -v jsonrpcserver [answer Y]
sudo pip3 install -v jsonrpcserver
tree -pu /usr/local/lib/python3.7/dist-packages/jsonrpcserver*
/usr/local/lib/python3.7/dist-packages/jsonrpcserver
├── [-rw-r--r-- root    ]  async_dispatcher.py
├── [-rw-r--r-- root    ]  dispatcher.py
├── [-rw-r--r-- root    ]  exceptions.py
├── [-rw-r--r-- root    ]  __init__.py
├── [-rw-r--r-- root    ]  log.py
├── [-rw-r--r-- root    ]  methods.py
├── [drwxr-sr-x root    ]  __pycache__
│   ├── [-rw-r--r-- root    ]  async_dispatcher.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  dispatcher.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  exceptions.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  __init__.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  log.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  methods.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  request.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  response.cpython-37.pyc
│   ├── [-rw-r--r-- root    ]  server.cpython-37.pyc
│   └── [-rw-r--r-- root    ]  status.cpython-37.pyc
├── [-rw-r--r-- root    ]  request.py
├── [-rw-r--r-- root    ]  request-schema.json
├── [-rw-r--r-- root    ]  response.py
├── [-rw-r--r-- root    ]  server.py
└── [-rw-r--r-- root    ]  status.py
/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0.egg-info
├── [-rw-r--r-- root    ]  dependency_links.txt
├── [-rw-r--r-- root    ]  installed-files.txt
├── [-rw-r--r-- root    ]  not-zip-safe
├── [-rw-r--r-- root    ]  PKG-INFO
├── [-rw-r--r-- root    ]  requires.txt
├── [-rw-r--r-- root    ]  SOURCES.txt
└── [-rw-r--r-- root    ]  top_level.txt

1 directory, 28 files
```
```
python3 -c 'from jsonrpcserver.exceptions import ApiError'
```
-> all ok

**Comment by nicola-lunghi** [comment]
(this happen to me because I'm using it wiith an embedded system and the build tool is calling python setup.py build install instead of pip) -> yocto

**Comment by bcb** [comment]
Interesting. I installed the same file (jsonrpcserver-4.1.0.tar.gz) using `python3 setup.py build install`, and it let me import ApiError successfully.

But that was in a virtualenv.

```
$ mkvirtualenv -p $(which python3.7) test-jsonrpcserver
Running virtualenv with interpreter /opt/local/bin/python3.7
Using base prefix '/opt/local/Library/Frameworks/Python.framework/Versions/3.7'
New python executable in /Users/beau/.virtualenvs/test-jsonrpcserver/bin/python3.7
Also creating executable in /Users/beau/.virtualenvs/test-jsonrpcserver/bin/python
Installing setuptools, pip, wheel...done.
(test-jsonrpcserver) /tmp
$ curl -O https://files.pythonhosted.org/packages/34/88/f224dcf47cbbbbea001c1bc981ba47dd39dc0ecd0cc98b7956d1839ade16/jsonrpcserver-4.1.0.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 14663  100 14663    0     0  14663      0  0:00:01 --:--:--  0:00:01 44033
(test-jsonrpcserver) /tmp
$ cd jsonrpcserver-4.1.0
(test-jsonrpcserver) /tmp/jsonrpcserver-4.1.0
$ python3 setup.py build install
running build
running build_py
creating build
creating build/lib
-- snip --
(test-jsonrpcserver) /tmp/jsonrpcserver-4.1.0
$ ls -l
total 32K
drwxr-xr-x  4 beau wheel  128 Jan  8 12:53 build/
drwxr-xr-x  3 beau wheel   96 Jan  8 12:53 dist/
drwxr-xr-x 13 beau wheel  416 Jan  6 13:19 jsonrpcserver/
drwxr-xr-x  8 beau wheel  256 Jan  8 12:53 jsonrpcserver.egg-info/
-rw-r--r--  1 beau wheel 4.4K Jan  6 13:07 CHANGELOG.md
-rw-------  1 beau wheel 1.1K Sep 19  2017 LICENSE
-rw-------  1 beau wheel   73 Sep 10 10:56 MANIFEST.in
-rw-r--r--  1 beau wheel 1.3K Jan  6 13:19 PKG-INFO
-rw-r--r--  1 beau wheel  583 Nov 15 11:09 README.md
-rw-r--r--  1 beau wheel   38 Jan  6 13:19 setup.cfg
-rw-r--r--  1 beau wheel 1.3K Jan  6 13:07 setup.py
(test-jsonrpcserver) /tmp/jsonrpcserver-4.1.0
$ python3
Python 3.7.0 (default, Jun 28 2018, 05:55:06)
[Clang 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Shell Enhancement module problem: No module named 'gnureadline'
>>> from jsonrpcserver.exceptions import ApiError
>>>
```

**Comment by bcb** [comment]
I'll make a new release with 644 on all the files.

**Comment by bcb** [comment]
Created a new 4.1.1 release, can you try that @nicola-lunghi ?

**Comment by andred** [comment]
Hi @bcb,

I have the same problem, and it's still the same with 4.1.1:

`tar tzvf jsonrpcserver-4.1.1.tar.gz
drwxr-xr-x beau/staff        0 2020-01-08 02:07 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/
-rw------- beau/staff     1239 2020-01-08 02:07 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/PKG-INFO
-rw-r--r-- beau/staff        1 2020-01-06 00:49 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/not-zip-safe
-rw------- beau/staff      566 2020-01-08 02:07 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/SOURCES.txt
-rw------- beau/staff      127 2020-01-08 02:07 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/requires.txt
-rw------- beau/staff       14 2020-01-08 02:07 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/top_level.txt
-rw------- beau/staff        1 2020-01-08 02:07 jsonrpcserver-4.1.1/jsonrpcserver.egg-info/dependency_links.txt
`

vs. previously on 4.0.5:

`tar tzvf jsonrpcserver-4.0.5.tar.gz
drwxr-xr-x beau/staff        0 2019-09-10 01:57 jsonrpcserver-4.0.5/jsonrpcserver.egg-info/
-rw-r--r-- beau/staff     1700 2019-09-10 01:57 jsonrpcserver-4.0.5/jsonrpcserver.egg-info/PKG-INFO
-rw-r--r-- beau/staff      502 2019-09-10 01:57 jsonrpcserver-4.0.5/jsonrpcserver.egg-info/SOURCES.txt
-rw-r--r-- beau/staff      127 2019-09-10 01:57 jsonrpcserver-4.0.5/jsonrpcserver.egg-info/requires.txt
-rw-r--r-- beau/staff       14 2019-09-10 01:57 jsonrpcserver-4.0.5/jsonrpcserver.egg-info/top_level.txt
-rw-r--r-- beau/staff        1 2019-09-10 01:57 jsonrpcserver-4.0.5/jsonrpcserver.egg-info/dependency_links.txt
`

**Comment by nicola-lunghi** [comment]
Hi!
> $ python3 setup.py build install
Here you are running this with the same user 

> (test-jsonrpcserver) /tmp/jsonrpcserver-4.1.0

the problem appear to be in the .egg directory

if you look up is

/usr/local/lib/python3.7/dist-packages/jsonrpcserver-4.1.0-py3.7**.egg**
├── [drwxr-sr-x root    ]  EGG-INFO
│   ├── [-rw------- root    ]  dependency_links.txt
│   ├── [-rw-r--r-- root    ]  not-zip-safe
│   ├── [-rw------- root    ]  PKG-INFO
│   ├── [-rw------- root    ]  requires.txt
│   ├── [-rw------- root    ]  SOURCES.txt
│   └── [-rw------- root    ]  top_level.txt

That has the permission problem, the normal directory is fine (why you are asking? I think that the egg is created by root and is copying the files as they are -> **only readable by root**

Your files are owned by the user beau that can access them so -> no error

Thank you!!!

**Comment by bcb** [comment]
Not sure why those files are getting `-rw----—`

I’ll take a closer look in the morning.

**Comment by bcb** [comment]
https://github.com/pypa/setuptools/issues/1328

**Comment by bcb** [comment]
This should be fixed in 4.1.2.

```sh
$ ls -l jsonrpcserver-4.1.2/jsonrpcserver.egg-info
total 24K
-rw-r--r-- 1 beau wheel 1.3K Jan  9 11:51 PKG-INFO
-rw-r--r-- 1 beau wheel  566 Jan  9 11:51 SOURCES.txt
-rw-r--r-- 1 beau wheel    1 Jan  9 11:51 dependency_links.txt
-rw-r--r-- 1 beau wheel    1 Jan  9 11:46 not-zip-safe
-rw-r--r-- 1 beau wheel  127 Jan  9 11:51 requires.txt
-rw-r--r-- 1 beau wheel   14 Jan  9 11:51 top_level.txt
```

This time I deleted my local jsonrpcserver.egg-info directory before creating the tar.gz. Seems to have worked, please let me know.

**Comment by bcb** [comment]
Closing as I believe it’s resolved.
